### PR TITLE
Read the Gemma4 merge dtype from the definition that reaches it

### PR DESCRIPTION
**Pull request by danielhanchen**
Follow-up to #1046, which fixed the red `Core (HF=latest + TRL=latest)` cell:

```
FAILED tests/test_gemma4_dtype_drift_guards.py::test_real_gemma4_image_and_video_merges_still_cast_dtype
```

#1046 replaced the substring search with a line search over assignments to
`<modality>_features`. That is correct on both transformers versions we run,
but it accepts a cast written anywhere in `modeling_gemma4.py`, which is looser
than the thing the guard exists to protect. Two files that are broken at the
merge still pass it:

```python
# a good cast that a later rebinding undoes
image_features = self.get_image_features(pixel_values).to(inputs_embeds.device, inputs_embeds.dtype)
image_features = postprocess(image_features).to(inputs_embeds.device)
inputs_embeds = inputs_embeds.masked_scatter(image_mask, image_features)
```

```python
# a cast that lives in a different method
def get_image_features(self, pixel_values, inputs_embeds):
    return self.vision_tower(pixel_values).to(inputs_embeds.device, inputs_embeds.dtype)

def forward(self, inputs_embeds):
    image_features = self.get_image_features(pixel_values, inputs_embeds)
    image_features = image_features.to(inputs_embeds.device)
    inputs_embeds = inputs_embeds.masked_scatter(image_mask, image_features)
```

### What this does

Reads the reaching definition instead of searching the file. Walk the AST, find
the `masked_scatter` that consumes `<modality>_features`, and check whether
`inputs_embeds.dtype` is applied in the scattered expression itself or in the
last assignment that reaches it. Both spellings upstream uses count, the
positional `.to(device, dtype)` overload and the keyword `dtype=`. A merge that
cannot be found at all is a failure rather than a silent pass.

That is what makes the later device-only rebinding win over the earlier good
cast, which is what actually runs, and what keeps a cast in an unrelated method
from standing in for one that is missing.

Robust across the versions we support rather than pinned to one upstream
spelling, which is what put the original guard on the floor:

| transformers | image | video | audio |
| --- | --- | --- | --- |
| 5.5.0 | `image_features = image_features.to(device, dtype)` | same shape | device only, the known upstream bug |
| 5.15.0 | cast folded into `torch.cat(image_features, dim=0).to(device, dtype)` | unchanged | now casts dtype at the merge |

Both shapes are dtype-aligned at the merge and both pass. Audio is unchanged
here: `test_real_gemma4_audio_merge_site_is_recognized` already accepts either
the device-only form or the fixed one, and upstream fixed it at 5.15.0.

`_MERGE_SHAPES` pins the behaviour on eight small sources and runs WITHOUT
gemma4 installed, which is the point: a guard that only runs on hosts with the
newest transformers is one nobody notices going vacuous. It covers the 5.5.0
shape, the 5.15.0 `torch.cat` shape, a cast written at the merge, the keyword
form, and the four ways this should go red, including the two cases above.

### Verified

- `tests/test_gemma4_dtype_drift_guards.py`: 35 passed on transformers 5.5.0
  and 35 passed on 5.15.0.
- Negative check on the real file: dropping `inputs_embeds.dtype` from the
  v5.15.0 `torch.cat` line makes the guard fail with `the image_features
  binding reaching the merge at line 2336 does not cast to inputs_embeds.dtype`,
  and it passes again once restored.
- Full `tests/` on transformers 5.15.0 + trl 1.9.2: 4470 passed, 297 skipped,
  9 failed. All 9 fail identically on unmodified main in this environment
  (they need `unsloth` installed and a generated trainer) and none touch gemma4.

Tests only, no runtime change.

### CODE REVIEW

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

File: tests/test_gemma4_dtype_drift_guards.py [hunk 1]
```diff
@@ -195,13 +196,113 @@ def _gemma4_modeling_source():
...
+    Covers both spellings upstream uses: the positional `.to(device, dtype)`
+    overload and the keyword `dtype=inputs_embeds.dtype` one.
+    """
+    for n in ast.walk(node):
+        if (
+            isinstance(n, ast.Attribute)
+            and n.attr == "dtype"
+            and isinstance(n.value, ast.Name)
+            and n.value.id == "inputs_embeds"
+        ):
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply tests/test_gemma4_dtype_drift_guards.py]: Correct, fixed in 806c2cc7. `_carries_inputs_embeds_dtype` matched the attribute anywhere in the subtree, so a mention that is not a cast argument counted. It is now `_call_aligns_dtype`, which only accepts `inputs_embeds.dtype` as a positional or `dtype=` argument of `.to(...)`. New fixture `dtype_mentioned_but_not_applied` uses your ternary and is red on the previous helper.

File: tests/test_gemma4_dtype_drift_guards.py [hunk 2]
```diff
@@ -195,13 +196,113 @@ def _gemma4_modeling_source():
...
+
+def _last_binding_before(tree, name, lineno):
+    """The last assignment to ``name`` that can reach line ``lineno``.
+
+    Textually last wins, which is what a straight-line merge block gives us.
+    Both `x = ...` and `x: T = ...` count, as does a tuple/starred target that
+    includes ``name``.
+    """
+    best = None
+    for node in ast.walk(tree):
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply tests/test_gemma4_dtype_drift_guards.py]: Correct, fixed in 806c2cc7. The candidate set is now the merge's own scope (`_scopes` plus `_walk_own_scope`, which does not descend into a nested def/class/lambda) intersected with the statement lists that enclose the merge (`_enclosing_blocks`), so a binding in a sibling branch is not a reaching definition. Rebindings inside an enclosing compound statement are folded per path, so an aligned `else` no longer covers a device-only `if`. New fixture `cast_in_untaken_branch` is red on the previous helper. I did not build a full dataflow over every path: the dominating-block fold is conservative in the failing direction, which is the safe one for a canary.

File: tests/test_gemma4_dtype_drift_guards.py [hunk 3]
```diff
@@ -195,13 +196,113 @@ def _gemma4_modeling_source():
...
+    """
+    name = f"{modality}_features"
+    found = []
+    for node in ast.walk(tree):
+        if not isinstance(node, ast.Call):
+            continue
+        func = node.func
+        if not (isinstance(func, ast.Attribute) and func.attr in ("masked_scatter", "masked_scatter_")):
+            continue
+        args = list(node.args) + [kw.value for kw in node.keywords if kw.arg in (None, "source")]
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply tests/test_gemma4_dtype_drift_guards.py]: Correct, fixed in 806c2cc7. The code scanned every positional argument while the docstring promised the source operand. `_feature_merges` now takes `node.args[1]`, or `source=`, which is torch's parameter name (`masked_scatter(mask, source)`; passing `tensor=` is a TypeError on torch 2.9.1). The mask is bool, so its dtype was never the thing that has to match. New fixture `cast_on_the_mask_only` is red on the previous helper.

File: tests/test_gemma4_dtype_drift_guards.py [hunk 4]
```diff
@@ -195,13 +196,113 @@ def _gemma4_modeling_source():
...
+        if isinstance(node, ast.Assign):
+            targets = node.targets
+        elif isinstance(node, ast.AnnAssign) and node.value is not None:
+            targets = [node.target]
+        else:
+            continue
+        binds = any(
+            isinstance(t, ast.Name) and t.id == name
+            for target in targets
+            for t in ast.walk(target)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply tests/test_gemma4_dtype_drift_guards.py]: Correct, fixed in 806c2cc7. `ast.walk(target)` matched a Name read inside a Subscript index, so a slice write was read as rebinding the whole tensor. Targets now go through `_bound_names`, which accepts Name plus tuple/list/starred positions and rejects Subscript and Attribute writes. New fixture `subscript_write_is_not_a_binding` is red on the previous helper.

File: tests/test_gemma4_dtype_drift_guards.py [hunk 5]
```diff
@@ -195,13 +196,113 @@ def _gemma4_modeling_source():
...
+    scattered VALUE (the second positional arg, or `source=`), which is the
+    tensor whose dtype has to line up with the destination.
+    """
+    name = f"{modality}_features"
+    found = []
+    for node in ast.walk(tree):
+        if not isinstance(node, ast.Call):
+            continue
+        func = node.func
+        if not (isinstance(func, ast.Attribute) and func.attr in ("masked_scatter", "masked_scatter_")):
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply tests/test_gemma4_dtype_drift_guards.py]: Correct, fixed in 806c2cc7. `_feature_merges` now requires the receiver to reference `inputs_embeds` before the call is treated as the protected merge. The failing direction you describe is the one I could build a realistic fixture for: `auxiliary_scatter_elsewhere` has a correct `inputs_embeds` merge followed by a `scratch.masked_scatter(...)` on a retyped tensor, and the previous helper went red on it.

File: tests/test_gemma4_dtype_drift_guards.py [hunk 6]
```diff
@@ -195,13 +196,113 @@ def _gemma4_modeling_source():
...
+    for call, arg in merges:
+        if _carries_inputs_embeds_dtype(arg):
+            continue
+        binding = _last_binding_before(tree, f"{modality}_features", call.lineno)
+        if binding is None:
+            return False, (
+                f"{modality}_features is scattered at line {call.lineno} with no "
+                f"assignment reaching it"
+            )
+        if not _carries_inputs_embeds_dtype(binding.value):
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply tests/test_gemma4_dtype_drift_guards.py]: Correct, fixed in 806c2cc7. `Tensor.to(device=None, dtype=None, ...)` documents that dtype is inferred from `self` when it is not given, so a device-only rebinding cannot drop the alignment. `_expr_alignment` now returns a third state for expressions that only move or reshape the tensor (`.to(<device>)`, indexing, and the shape/placement methods in `_DTYPE_PRESERVING_METHODS`), and `_reaching_alignment` lets those inherit the running verdict instead of resetting it. New fixture `device_move_after_cast` is red on the previous helper. Source: https://docs.pytorch.org/docs/stable/generated/torch.Tensor.to.html

File: tests/test_gemma4_dtype_drift_guards.py [hunk 7]
```diff
@@ -195,13 +196,113 @@ def _gemma4_modeling_source():
...
+    Returns (ok, detail) so a failure can say what it saw.
+    """
+    tree = ast.parse(src)
+    merges = _merge_calls(tree, modality)
+    if not merges:
+        return False, f"no masked_scatter consuming {modality}_features"
+    for call, arg in merges:
+        if _carries_inputs_embeds_dtype(arg):
+            continue
+        binding = _last_binding_before(tree, f"{modality}_features", call.lineno)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply tests/test_gemma4_dtype_drift_guards.py]: Correct, fixed in 806c2cc7, and it is the necessary counterpart to the device-move item: the binding is now consulted ONLY when the scattered expression is demonstrably dtype-preserving. Anything else, including `.float()`, `.to(torch.float32)` and `.to(other.dtype)`, is unaligned outright rather than falling back to an earlier good cast. New fixture `source_overrides_the_dtype` is red on the previous helper.

File: tests/test_gemma4_dtype_drift_guards.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (danielhanchen) [reply tests/test_gemma4_dtype_drift_guards.py]: Correct, fixed in 806c2cc7. `.to(other)` is documented as returning a tensor with the same dtype and device as `other`, and `.type_as(other)` gives the dtype; both are now accepted by `_call_aligns_dtype`. Confirmed on torch 2.9.1: `float32_tensor.to(bf16_ref).dtype` and `.type_as(bf16_ref).dtype` are both bfloat16, and there is no `other=` keyword, so only the first positional argument is read. New fixtures `tensor_overload` and `type_as` are red on the previous helper. Source: https://docs.pytorch.org/docs/stable/generated/torch.Tensor.to.html

**Comment by danielhanchen** [comment]
Overlap note for whoever merges this: #1037 rewrites the same helper in the same file and is further along. It confines the trace to `Gemma4Model.forward`, checks the receiver, reads only `args[1]`, excludes subscript targets, recognises `.to(other)`, and adds a dominator analysis plus a merge-result-is-used check that this PR does not have. The two will conflict textually in `tests/test_gemma4_dtype_drift_guards.py`.

The one thing this PR covers that #1037 does not is the device-only rebinding: `image_features = image_features.to(inputs_embeds.device)` after an earlier `inputs_embeds.dtype` cast preserves the dtype (`.to()` infers dtype from self when none is given), and #1037 reads it as dropping the alignment. If #1037 lands first, that one case is worth carrying over and the rest of this PR can go.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
Closing this as superseded by #1037, which merged a strictly stronger version of the same guard.

I opened this to replace the regex-based drift check in `tests/test_gemma4_dtype_drift_guards.py` with an AST one. #1037 landed the production fix in `unsloth_zoo/temporary_patches/gemma4.py` and rewrote that same test file at the same time, and its version is the more complete of the two:

| | this PR | #1037, now on main |
|---|---|---|
| test functions | 12 | 37 |
| `.to(dtype)` spellings recognised | explicit `features.to(inputs_embeds.device, inputs_embeds.dtype)` | that, plus the tensor overload `features.to(inputs_embeds)` |
| dominance / reachability of the assignment that feeds the merge | not modelled | modelled |

Rebasing this onto the current main conflicts throughout that file, and resolving it would mean reintroducing the narrower checks. Every capability this branch added is present on main: `_is_dtype_bearing` and `_is_device_expr` verbatim, and this branch's `_is_embeds_dtype` and `_mentions` are subsumed by main's `_is_inputs_embeds` and `_casts_to_inputs_embeds_dtype`.

Nothing is lost by closing. If I find a case main's version misses, it belongs in a fresh PR against the current file rather than in this branch.
